[ci skip] Introduce a `ignore-rev` file for better blaming experience

Earlopain opened the pull request:
https://github.com/rails/rails/commit/80e66cc4d90bf8c15d1a5f6e3152e90147f00772: 11,638 additions and 11,661 deletions

There are some commits in rails that touch a large amount of files and are cosmetic only.

This file will do two things:
* Blaming through the GitHub UI will not blame these commits
* Locally you can use it with `--ignore-revs-file`

https://docs.github.com/en/repositories/working-with-files/using-files/viewing-a-file#ignore-commits-in-the-blame-view https://git-scm.com/docs/git-blame#Documentation/git-blame.txt---ignore-revs-fileltfilegt

There surely are more commits worthy of being included here but I just ran into this one once again and would like to stop doing that.

This comment gets linked to quite often https://github.com/rails/rails/pull/13771#issuecomment-32746700 and while there seldom are large cosmetic changes, they do exist. Let's make the ones that did happen a bit less painful.

-----

I also remember bumping against the string literal changes a few times (https://github.com/rails/rails/commit/35b3de8021e68649cac963bd82a74cc75d1f60f0 et al) , added those to the list as well.

yahonda commented:
Hi, I prefer not to maintain this file in the Rails repository, as opinions vary on which commits to ignore, and I don’t want to keep updating it for future commits.

Earlopain commented:
I don't think that is a particularly good argument. Rails doesn't even allow cosmetic commits like that and they are rejected when attempted by individual contributors. Here they were made by a core team member and huge stylistic changes like that are incredibly rare.

Blame being messed up by such commits is one reason not to do them and I would find it sad when this would be rejected even though a solution to make it less painful exists.

For example, take a look at the file in the ruby repo:  https://github.com/ruby/ruby/blob/master/.git-blame-ignore-revs
It is also a huge repo with many commits and they only list 17 commits. There are already so many in this PR because the `'` to `"` conversion was made in commits per directory instead of in one go.
Another example from rust, https://github.com/rust-lang/rust/blob/master/.git-blame-ignore-revs, equally small as for ruby.

matthewd commented:
I think this seems reasonable. I'm sad it means adding yet another top-level file, but more cooperative blame behaviour is probably worth that price.

As for the potential for ongoing maintenance... I just did a quick eyeball of all commits that touched 100+ files, and then all commits with 100+ insertions _and_ deletions back about a decade. The list I came up with is nontrivial (and not exhaustive: I ignored some, and doubtless missed others), but still pretty manageable. 

Between my candidates and the original list in this PR, we're at 79 commits across 10+ years, even before accounting for the several bursts (like the string literals set) that occurred together.

There's always a risk that people will propose lots of inappropriate changes, but I think we can manage that. (I'm not convinced we should add my full list, tbh: should we focus only on commits that are known by observation to show up in blames frequently [like the PR's current ones], and/or generally filter to commits that would otherwise be blameable for lines currently present in HEAD?)

---

* d6bf4de7dc Transform actionpack documentation to Markdown
* 19f8ab2e7d [Tests only] Enable `Minitest/AssertPredicate` rule
* 18707ab17f Standardize nodoc comments
* aa3dcabd87 Add `Style/RedundantFreeze` to remove redudant `.freeze`
* 1b86d90136 Enable `Performance/UnfreezeString` cop
* 4c0a3d4880 Arel: rubocop -a
* 94333a4c31 Use assert_predicate and assert_not_predicate
* 40bdbce191 Define path with __dir__
* fe1f4b2ad5 Add more rubocop rules about whitespaces
* 55f9b8129a Add three new rubocop rules
* b326e82dc0 applies remaining conventions across the project
* 411ccbdab2 remove redundant curlies from hash arguments
* 5c315a8fa6 modernizes hash syntax in activesupport
* d22e522179 modernizes hash syntax in activerecord
* 63fff600ac modernizes hash syntax in actionview
* 5b6eb1d58b modernizes hash syntax in actionpack
* e6ab70c439 applies new string literal convention to the rest of the project
* d1374f99bf Pass symbol as an argument instead of a block
* b95d6e84b0 Deletes trailing whitespaces (over text files only find * -type f -exec sed 's/[ \t]*$//' -i {} \;)
* b451de0d6d Deletes trailing whitespaces (over text files only find * -type f -exec sed 's/[ \t]*$//' -i {} \;)
* dd89f600f7 :link: Remove RDoc auto-link from Rails module everywhere
* e3c73fd183 Replace all occurrences of '<tt>(\w+::\w+)</tt>' with '+$1+'
* e37adfed4e Add Oxford commas [ci-skip]
* 258670244d Added a lot of Oxford commas
* a1ac18671a Replace `assert !` with `assert_not`
* 89bcca59e9 Remove usage of strip_heredoc in the framework in favor of <<~
* 82c39e1a0b Use assert_empty and assert_not_empty
* 0d50cae996 Use respond_to test helpers
* d5ea33b9a0 [Active Record] require => require_relative
* 6a728491b6 [Railties] require_relative => require
* 9d4f79d3d3 [Active Record] require => require_relative
* 618268b4b9 [Railties] require => require_relative
* 3d453b409d Make ActionView frozen string literal friendly.
* b70fc698e1 Reduce string objects by using \ instead of + or << for concatenating strings
* 010e246756 Fix Rubocop violations and fix documentation visibility
* e8ba0c0f21 "Use assert_nil if expecting nil. This will fail in minitest 6."
* 2bdc4ddc06 Code styling cleanup
* fbccae4d19 Fix Remaining Case-In-Assignment Statement Formatting
* a9dc45459a code gardening: removes redundant selfs
* 74497eabd5 Revert "Revert "Eliminate the EventMachine dependency""
* 9ea7aa84d1 Revert "Eliminate the EventMachine dependency"
* b5bbdbd3bc Revert "Add prepared statements support for `Mysql2Adapter`"
* 38746d085b Add prepared statements support for `Mysql2Adapter`
* 428d47adfe applies new doc guidelines to Active Record.
* eebb9ddf9b Convert ActiveModel to 1.9 hash syntax.
* eb493f5ac8 update AD::Journey to follow Rails coding conventions
* 3c580182ff 1.9 hash syntax changes
* d20a52930a 1.9 hash syntax changes to docs
* be4a4cd38f Hash Syntax to 1.9 related changes
* 5ad7f8ab41 Use Ruby 1.9 Hash syntax in railties
* a2c8438854 1.9 hash syntax changes.
* 96f290eac0 Update actionmailer with new hash syntax.
* d76ef6fa6c Update some code styles.

(Produced by searching and scanning the output of `git log --oneline --shortstat`)

Earlopain committed:
Introduce a `ignore-rev` file for better blaming experience

https://github.com/rails/rails/commit/80e66cc4d90bf8c15d1a5f6e3152e90147f00772: 11,638 additions and 11,661 deletions

There are some commits in rails that touch a large amount of files
and are cosmetic only.

This file will do two things:
* Blaming through the GitHub UI will not blame these commits
* Locally you can use it with `--ignore-revs-file`

https://docs.github.com/en/repositories/working-with-files/using-files/viewing-a-file#ignore-commits-in-the-blame-view
https://git-scm.com/docs/git-blame#Documentation/git-blame.txt---ignore-revs-fileltfilegt

There surely are more commits worthy of being included here but I just ran into this one
once again and would like to not do that again.

This comment gets linked to quite often https://github.com/rails/rails/pull/13771#issuecomment-32746700
and while there seldom are large cosmetic changes, they do exist.
Let's make the ones that did happen a bit less painful.

Earlopain commented:
I added some of the commits you dug up. I looked at them and excluded commits like 3d453b409d037a056d0bcd636a2e020cc7cef4a8 which have a concrete reason for being done. Others were reverts like d5ea33b9a003d32ebaeae7ad89c479c04f790642 which I also believe are valuable.

> should we focus only on commits that are known by observation to show up in blames frequently [like the PR's current ones], and/or generally filter to commits that would otherwise be blameable for lines currently present in HEAD?

Rails has a long history, I often have to go back years to understand why something happenend so excluding even these "ancient" commits still has value. I would not go out of my way to find these though. With this initial list it will already be much improved.

If this list is too long now, I wouldn't be opposed to going back to the previous version (and maybe adding the `modernizes hash syntax *` commits). I believe that list will still have a big impact.

---------

Also added the comment about cosmetic changes from the guide/contributing. Seems relevant.

byroot commented:
As someone who spend a lot of time in git blame I agree this is valuable.

This list will never be exhaustive nor perfect, but filtering the worst offenders is good. Let's start with that.
